File: test/bench/heap/heap.js
class Node {
    left
    right
    parent
    value

    constructor(left, right, parent, value) {
        this.left = left
        this.right = right
        this.parent = parent
        this.value = value
    }

    getLeftmost() {
        if (this.left == null) {
            return this
        }
        return this.left.getLeftmost()
    }

    getRightmost() {
        if (this.right == null) {
            return this
        }
        return this.right.getRightmost()
    }

    getLeftSibling() {
        if (this.parent.right == this) {
            return this.parent.left
        } else {
            return this.parent.getLeftSibling2(1)
        }
    }

    getLeftSibling2(height) {
        if (this.parent.right == this) {
            return this.parent.left.getRightN(height)
        } else {
            return this.parent.getLeftSibling2(height + 1)
        }
    }

    getRightSibling() {
        if (this.parent.left == this) {
            return this.parent.right
        } else {
            return this.parent.getRightSibling2(1)
        }
    }

    getRightSibling2(height) {
        if (this.parent.left == this) {
            return this.parent.right.getLeftN(height)
        } else {
            return this.parent.getRightSibling2(height + 1)
        }
    }

    getRightN(n) {
        if (n == 1) {
            return this.right
        } else {
            return this.right.getRightN(n - 1)
        }
    }

    getLeftN(n) {
        if (n == 1) {
            return this.left
        } else {
            return this.left.getLeftN(n - 1)
        }
    }
}

class Heap {
    root
    size
    last

    constructor(root, size, last) {
        this.root = root
        this.size = size
        this.last = last
    }
    
    insert(value) {
        if (this.root == null) {
            this.root = new Node(null, null, null, value)
            this.last = this.root
            this.size = 1
            return
        }
        if (((this.size + 1) & this.size) == 0) {
            // Insert at left most node.
            var parent = this.root.getLeftmost()
            var newN = new Node(null, null, parent, value)
            parent.left = newN
            this.last = newN
        } else {
            // Insert after last node.
            if (this.size % 2 == 0) {
                var newN = new Node(null, null, this.last.parent, value)
                this.last.parent.right = newN
                this.last = newN
            } else {
                var sibling = this.last.parent.getRightSibling()
                var newN = new Node(null, null, sibling, value)
                sibling.left = newN
                this.last = newN
            }
        }
        this.size = this.size + 1
        this.siftUp(newN)
    }

    swapUp(node) {
        if (this.last == node) {
            this.last = node.parent
        }
        var parentSave = node.parent
        var parentLeft = parentSave.left
        var parentRight = parentSave.right
        var parentParent = parentSave.parent

        parentSave.left = node.left
        if (node.left != null) {
            node.left.parent = parentSave
        }
        parentSave.right = node.right
        if (node.right != null) {
            node.right.parent = parentSave
        }
        parentSave.parent = node

        if (parentLeft == node) {
            node.left = parentSave
            node.right = parentRight
            if (parentRight != null) {
                parentRight.parent = node
            }
        } else {
            node.left = parentLeft
            node.right = parentSave
            if (parentLeft != null) {
                parentLeft.parent = node
            }
        }
        node.parent = parentParent

        if (parentParent == null) {
            this.root = node
        } else {
            if (parentParent.left == parentSave) {
                parentParent.left = node
            } else {
                parentParent.right = node
            }
        }
    }

    siftUp(node) {
        if (node.parent == null) {
            return
        }
        if (node.value > node.parent.value) {
            this.swapUp(node)
            if (this.root != node) {
                this.siftUp(node)
            }
        }
    }

    siftDown(node) {
        if ((node.left != null) && node.left.value > node.value) {
            this.swapUp(node.left)
            this.siftDown(node)
        } else if ((node.right != null) && node.right.value > node.value) {
            this.swapUp(node.right)
            this.siftDown(node)
        }
    }

    popTop() {
        if (this.size == 1) {
            var res = this.root
            this.root = null
            this.last = null
            this.size = 0
            return res
        }

        if (this.size <= 3) {
            var top = this.root
            this.swapUp(this.last)
            top.parent = null
            if (this.size == 3) {
                this.root.right = null
                this.last = this.root.left
            } else {
                this.root.left = null
                this.last = this.root
            }
            this.size = this.size - 1
            return top
        }

        if ((this.size & (this.size - 1)) == 0) {
            var newLast = this.root.getRightmost()
        } else {
            var newLast = this.last.getLeftSibling()
        }

        // Move last to top and remove last.
        this.root.left.parent = this.last
        this.last.left = this.root.left
        this.root.right.parent = this.last
        this.last.right = this.root.right

        if (this.last.parent.left == this.last) {
            this.last.parent.left = null
        } else {
            this.last.parent.right = null
        }
        this.last.parent = null

        var top = this.root
        this.root = this.last
        this.last = newLast
        this.size = this.size - 1

        this.siftDown(this.root)
        return top
    }
}

var h = new Heap(null, 0, null)
for (var i = 1; i < 20000; i++) {
    h.insert(i) 
}

var sum = 0
for (var i = 1; i < 20000; i++) {
    sum = sum + h.popTop().value
}

console.log(sum)

File: test/bench/heap/heap.lua
Node = {}
Node.__index = Node

function Node:new(left, right, parent, value)
    local self = setmetatable({}, Node)
    self.left = left
    self.right = right
    self.parent = parent
    self.value = value
    return self
end

function Node:getLeftmost()
    if (self.left == null) then
        return self
    end
    return self.left:getLeftmost()
end

function Node:getRightmost()
    if (self.right == null) then
        return self
    end
    return self.right:getRightmost()
end

function Node:getLeftSibling()
    if (self.parent.right == self) then
        return self.parent.left
    else
        return self.parent:getLeftSibling2(1)
    end
end

function Node:getLeftSibling2(height)
    if (self.parent.right == self) then
        return self.parent.left:getRightN(height)
    else
        return self.parent:getLeftSibling2(height + 1)
    end
end

function Node:getRightSibling()
    if (self.parent.left == self) then
        return self.parent.right
    else
        return self.parent:getRightSibling2(1)
    end
end

function Node:getRightSibling2(height)
    if (self.parent.left == self) then
        return self.parent.right:getLeftN(height)
    else
        return self.parent:getRightSibling2(height + 1)
    end
end

function Node:getRightN(n) 
    if (n == 1) then
        return self.right
    else
        return self.right:getRightN(n - 1)
    end
end

function Node:getLeftN(n)
    if (n == 1) then
        return self.left
    else
        return self.left:getLeftN(n - 1)
    end
end

Heap = {}
Heap.__index = Heap

function Heap:new(root, size, last)
    local self = setmetatable({}, Heap)
    self.root = root
    self.size = size
    self.last = last
    return self
end
    
function Heap:insert(value) 
    if (self.root == null) then
        self.root = Node:new(null, null, null, value)
        self.last = self.root
        self.size = 1
        return
    end
    local newN
    if ((self.size + 1) & self.size == 0) then
        -- Insert at left most node.
        local parent = self.root:getLeftmost()
        newN = Node:new(null, null, parent, value)
        parent.left = newN
        self.last = newN
    else
        -- Insert after last node.
        if (self.size % 2 == 0) then
            newN = Node:new(null, null, self.last.parent, value)
            self.last.parent.right = newN
            self.last = newN
        else
            local sibling = self.last.parent:getRightSibling()
            newN = Node:new(null, null, sibling, value)
            sibling.left = newN
            self.last = newN
        end
    end
    self.size = self.size + 1
    self:siftUp(newN)
end

function Heap:swapUp(node)
    if (self.last == node) then
        self.last = node.parent
    end
    local parentSave = node.parent
    local parentLeft = parentSave.left
    local parentRight = parentSave.right
    local parentParent = parentSave.parent

    parentSave.left = node.left
    if (node.left ~= null) then
        node.left.parent = parentSave
    end
    parentSave.right = node.right
    if (node.right ~= null) then
        node.right.parent = parentSave
    end
    parentSave.parent = node

    if (parentLeft == node) then
        node.left = parentSave
        node.right = parentRight
        if (parentRight ~= null) then
            parentRight.parent = node
        end
    else
        node.left = parentLeft
        node.right = parentSave
        if (parentLeft ~= null) then
            parentLeft.parent = node
        end
    end
    node.parent = parentParent

    if (parentParent == null) then
        self.root = node
    else
        if (parentParent.left == parentSave) then
            parentParent.left = node
        else
            parentParent.right = node
        end
    end
end

function Heap:siftUp(node)
    if (node.parent == null) then
        return
    end
    if (node.value > node.parent.value) then
        self:swapUp(node)
        if (self.root ~= node) then
            self:siftUp(node)
        end
    end
end

function Heap:siftDown(node)
    if ((node.left ~= null) and node.left.value > node.value) then
        self:swapUp(node.left)
        self:siftDown(node)
    elseif ((node.right ~= null) and node.right.value > node.value) then
        self:swapUp(node.right)
        self:siftDown(node)
    end
end

function Heap:popTop()
    if (self.size == 1) then
        local res = self.root
        self.root = null
        self.last = null
        self.size = 0
        return res
    end

    if (self.size <= 3) then
        local top = self.root
        self:swapUp(self.last)
        top.parent = null
        if (self.size == 3) then
            self.root.right = null
            self.last = self.root.left
        else
            self.root.left = null
            self.last = self.root
        end
        self.size = self.size - 1
        return top
    end

    local newLast
    if (self.size & (self.size - 1) == 0) then
        newLast = self.root:getRightmost()
    else
        newLast = self.last:getLeftSibling()
    end

    -- Move last to top and remove last.
    self.root.left.parent = self.last
    self.last.left = self.root.left
    self.root.right.parent = self.last
    self.last.right = self.root.right

    if (self.last.parent.left == self.last) then
        self.last.parent.left = null
    else
        self.last.parent.right = null
    end
    self.last.parent = null

    local top = self.root
    self.root = self.last
    self.last = newLast
    self.size = self.size - 1

    self:siftDown(self.root)
    return top
end

local h = Heap:new(null, 0, null)
for i = 1, 20000-1 do
    h:insert(i) 
end

local sum = 0
for i = 1, 20000-1 do
    sum = sum + h:popTop().value
end

io.write(sum .. "\n")